staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered_test.go

```diff
@@ -0,0 +1,295 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package buffered
+
+import (
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	auditinternal "k8s.io/apiserver/pkg/apis/audit"
+	"k8s.io/apiserver/pkg/audit"
+)
+
+func newBufferedBackend() (*bufferedBackend, *fakeBackend) {
+	fakeBackend := newFakeBackend()
+	backend := NewBackend(fakeBackend)
+	return backend.(*bufferedBackend), fakeBackend
+}
+
+func newBufferedBackendWithWaitGroup() (*bufferedBackend, *fakeBackend) {
+	fakeBackend := newFakeBackend()
+	fakeBackend.wg = new(sync.WaitGroup)
+	backend := NewBackend(fakeBackend)
+	return backend.(*bufferedBackend), fakeBackend
+}
+
+func newBufferedBackendWithChannel() (*bufferedBackend, *fakeBackend) {
+	fakeBackend := newFakeBackend()
+	fakeBackend.ch = make(chan struct{})
+	backend := NewBackend(fakeBackend)
+	return backend.(*bufferedBackend), fakeBackend
+}
+
+// waitForEmptyBuffer indicates when the sendBatchEvents method has read from the
+// existing buffer. This lets test coordinate closing a timer and stop channel
+// until the for loop has read from the buffer.
+func waitForEmptyBuffer(b *bufferedBackend) {
+	for len(b.buffer) != 0 {
+		time.Sleep(time.Millisecond)
+	}
+}
+
+func TestBatchMaxEvents(t *testing.T) {
```

Comment:
If you just add `chan []*auditinternal.Event` to the fakeBackend, you can do the following:

```go
fakeBackend := newFakeBackend()
bufferedBackend := NewBackend(fakeBackend)

bufferedBackend.Run(wait.NeverStop)
bufferedBackend.ProcessEvents(events...)

batch := getBatchWithTimeout(t, fakeBackend)
require.Equal(t, defaultBatchMaxSize, len(batch), ...)
batch = getBatchWithTimeout(t, fakeBackend)
require.Equal(t, 10, len(batch), ...)
```